client/.eslintrc.js

```diff
@@ -26,6 +26,15 @@ module.exports = {
 			},
 		],
 		'jest/no-mocks-import': 'off',
+		'no-restricted-syntax': [
+			'error',
+			{
+				selector:
+					'CallExpression[callee.name="useSelector"][arguments.0.body.type="ObjectExpression"]',
```

Comment:
Unfortunately, this won't catch:

- Objects returned in a block statement
- Arrays - either directly or in a block statement

So, how about we modify to catch more:

```suggestion
					'CallExpression[callee.name="useSelector"] ArrowFunctionExpression > :matches(ObjectExpression, ArrayExpression), CallExpression[callee.name="useSelector"] ArrowFunctionExpression > BlockStatement ReturnStatement > :matches(ObjectExpression, ArrayExpression)',
```

I've tried this out with:

1. Object return shorthand

```js
const { sitesAsLandingPage } = useSelector( ( state ) => ( {
	sitesAsLandingPage: getPreference( state, 'sites-landing-page' )?.useSitesAsLandingPage,
} ) );
```

2. Object returned in a block statement

```js
const { readerAsLandingPage } = useSelector( ( state ) => {
	const a = 1;
	return {
		readerAsLandingPage: a,
	};
} );
```

3. Array return shorthand

```js
const [ asd ] = useSelector( ( state ) => [ state ] );
```

4. Array returned in a block statement

```js
const [ x ] = useSelector( ( state ) => {
	const a = 1;
	return [ a, state ];
} );
```

Reply:
Good call! I expected this was imperfect, but didn't think the other patterns were that common. Based on the additional findings, turns out I'm wrong! 😄 

Updated in 13d42aba9e1 with a small adjustment to use stricter child selectors, since otherwise it could incorrectly identify some return statements inside nested functions that aren't a problem.

Contrived example:

```ts
const { a } = useSelector( ( state ) => {
	[].map( () => { return []; } );
	return { a: 1 };
} );
```